modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/pagemem/SpeedBasedMemoryConsumptionThrottlingStrategy.java

```diff
@@ -367,11 +325,30 @@ private long calcSpeedToMarkAllSpaceTillEndOfCp(double dirtyPagesRatio, long don
         if (dirtyPagesRatio >= MAX_DIRTY_PAGES)
             return 0;
 
-        double remainedClearPages = (MAX_DIRTY_PAGES - dirtyPagesRatio) * totalPages;
+        // IDEA: here, when calculating the count of clean pages, it includes the pages under checkpoint. It is kinda
+        // legal because they can be written (using the Checkpoint Buffer to make a copy of the value to be
+        // checkpointed), but the CP Buffer is usually not too big, and if it gets nearly filled, writes become
+        // throttled really hard by exponential throttler. Maybe we should subtract the number of not-yet-written-by-CP
+        // pages from the count of clean pages? In such a case, we would lessen the risk of CP Buffer-caused throttling.
+        double remainedCleanPages = (MAX_DIRTY_PAGES - dirtyPagesRatio) * pageMemTotalPages();
+
+        double secondsTillCPEnd = 1.0 * (cpTotalPages - donePages) / avgCpWriteSpeed;
+
+        return (long)(remainedCleanPages / secondsTillCPEnd);
+    }
+
+    /** Returns total number of pages storable in page memory. */
+    private long pageMemTotalPages() {
+        long currentTotalPages = pageMemTotalPages;
+
+        if (currentTotalPages == 0) {
```

Comment:
When is it 0 I wonder?

Reply:
It is 0 in the very beginning, before it was written